### PR TITLE
Refactor Builder to lookup for builder before build_extension is called.

### PR DESCRIPTION
**Pull request by simi**
followup to https://github.com/rubygems/rubygems/pull/2850 @duckinator @bronzdoc 

It is possible to remove `@ran_rake` at all. If we find builder before we try to construct `extension_dir` in `build_extension` method, it will raise on invalid names like empty string or nil.

### CODE REVIEW

File: lib/rubygems/ext/builder.rb [hunk 1]
```diff
@@ -123,12 +121,11 @@ def builder_for(extension) # :nodoc:
     when /configure/ then
       Gem::Ext::ConfigureBuilder
     when /rakefile/i, /mkrf_conf/i then
-      @ran_rake = true
       Gem::Ext::RakeBuilder
     when /CMakeLists.txt/ then
       Gem::Ext::CmakeBuilder
     else
-      extension_dir = File.join @gem_dir, File.dirname(extension)
+      extension_dir = File.join @gem_dir, File.dirname(extension.to_s)
```
Comment (simi) [inline lib/rubygems/ext/builder.rb]: This is the only place we need valid string.

File: lib/rubygems/ext/builder.rb [hunk 2]
```diff
@@ -221,9 +203,10 @@ def build_extensions
     FileUtils.rm_f @spec.gem_build_complete_path
 
     @spec.extensions.each do |extension|
-      break if @ran_rake
+      builder = builder_for extension
+      build_extension extension, builder, dest_path
 
-      build_extension extension, dest_path
+      break if builder.is_a? Gem::Ext::RakeBuilder
```
Comment (simi) [inline lib/rubygems/ext/builder.rb]: Code breaks after rake builder is triggered for some reason. I have no idea why this is needed. All spec passes also without this. It was added at https://github.com/rubygems/rubygems/commit/20cdf22b844a63d163cc80eff819b99330bd6fa1.

In theory if you have multiple extensions defined in gemspec, if rake one is specified, no other extension one after rake will be build.

:thinking: I don't see any reason for this. But for now I have kept this to not modify any behaviour.

Comment (MSP-Greg) [reply lib/rubygems/ext/builder.rb]: Haven't been following that closely, but...

> In theory if you have multiple extensions defined in gemspec

An example would be EventMachine, two extconf.rb files, and two *.so files compiled.

Comment (simi) [reply lib/rubygems/ext/builder.rb]: As I see https://github.com/eventmachine/eventmachine/blob/d761d2a6916b854b5d62aeaa4775be359e9cf6ce/eventmachine.gemspec#L16 there are two extensions defined, but both using `extconf` builder. But if first one will use rake builder, the second one will not be built and I have no idea what's the reason for this behaviour.

**Comment by bronzdoc** [comment]
This PR would be easier to understand after we merge https://github.com/rubygems/rubygems/pull/2849?

what you think @simi @MSP-Greg ?

**Comment by MSP-Greg** [comment]
@bronzdoc

Earlier today I saw the conversation, and it seemed to mention issues with multiple extensions.  I recalled that EM had two, so I mentioned it.  I've never had issues with building on Windows (at least related to RG), so I've never looked at the RG code.  Hence, right now I have no opinion.

I'll have a look at the code tomorrow (-0500 from GMT).

**Comment by simi** [comment]
@bronzdoc #2849 does a little another approach, but I can combine the best of both PRs into this one. I have not noticed that one before, I'm sorry.

@MSP-Greg I'm just sharing my confusion about this behaviour, anyway this PR doesn't change it.

**Comment by duckinator** [comment]
This looks like it should be good, but I agree that rebasing it will make it easier to understand (just because the diff will be smaller), so I'd be a bit more confident in my assessment at that point.

@simi when you have a moment, could you rebase this? (since #2849 was merged.)

Thanks for the PR, by the way!

**Comment by bronzdoc** [comment]
Ah I missed your comment @simi 😞

**Comment by bronzdoc** [comment]
@simi do you want me to pick this up?

**Pull request closed without merging**